crates/fsrt/src/main.rs

```diff
@@ -99,7 +105,90 @@ impl fmt::Display for Error {
     }
 }
 
-impl std::error::Error for Error {}
+fn check_graphql_and_perms(val: &Value) -> Vec<&str> {
+    let mut operations = vec![];
+    match val {
+        Value::Const(Const::Literal(s)) => operations.extend_from_slice(&parse_graphql(s)),
+        Value::Phi(vals) => vals.iter().for_each(|val| match val {
+            Const::Literal(s) => operations.extend_from_slice(&parse_graphql(s)),
+        }),
+        _ => {}
+    }
+    // TODO : Build out permission resolver here
+
+    let permissions_resolver: HashMap<(String, String), &str> = [(
+        ("compass".into(), "searchTeams".into()),
+        "read:component:compass",
+    )]
+    .into_iter()
+    .collect();
+
+    operations
+        .iter()
+        .filter_map(|f| permissions_resolver.get(f).copied())
+        .collect()
+}
+
+fn parse_graphql(s: &str) -> Vec<(String, String)> {
+    let mut operations = vec![];
+    let mut fragments: HashMap<&str, graphql_parser::query::SelectionSet<'_, &str>> =
+        HashMap::new();
+
+    if let std::result::Result::Ok(doc) = parse_query::<&str>(s) {
+        doc.definitions.clone().into_iter().for_each(|d| {
+            if let Definition::Fragment(fragment_def) = d {
+                fragments.insert(fragment_def.name, fragment_def.selection_set);
+            }
+        });
+
+        doc.definitions.into_iter().for_each(|operation| {
+            if let Definition::Operation(op) = operation {
+                let selection_set = match op {
+                    OperationDefinition::Mutation(mutation) => Some(mutation.selection_set),
+                    OperationDefinition::Query(query) => Some(query.selection_set),
+                    OperationDefinition::SelectionSet(set) => Some(set),
+                    _ => None,
+                };
+                if let Some(set) = selection_set {
+                    set.items.into_iter().for_each(|f| {
+                        operations.extend_from_slice(&get_field_and_operation(f, &fragments));
+                    })
+                }
+            }
+        })
+    }
+
+    operations
+}
+
+fn get_field_and_operation<'a, 'b>(
+    selection: Selection<'a, &'a str>,
+    fragments: &'b HashMap<&'b str, SelectionSet<'a, &'a str>>,
+) -> Vec<(String, String)> {
+    let mut vec = vec![];
+
+    if let Selection::Field(mut type_field) = selection {
+        type_field.selection_set.items.clone().iter().for_each(|f| {
```

Comment:
```suggestion
        type_field.selection_set.items.iter().for_each(|f| {
```

Reply:
There is a mutable reference within the iteration (when we replace the FragmentSpread with the Fragment's fields), so I think this needs to be cloned here. Unless I am missing something, which I usually am 🙅

Reply:
Wait, can't you just do it in one pass by iterating over the fields of the fragment spread?

Also, looking at this, you don't have to return `String`s and could just use the `&str`s directly by returning `Vec<(&'a str, &'a str)>`.

Reply:
You are right. Only one pass was needed. I removed the second pass. Thanks as always, Josh!